### PR TITLE
feat: add design document for hbar allowance/approval

### PR DESCRIPTION
**Pull request by lukelee-sl**
**Description**:
Create a design document that describes adding functionality to smart contracts which allow for granting allowance for hbar to a spender account.

- [x] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (Nana-EC) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:71]:
```suggestion
- Test that a contract calling the `hbarApprove` function without the correct contract key will fail.
```

**Review by Nana-EC — COMMENTED** [review]
LG so far, let's wait on the HIP to make more progress via reviews and get merged before merging this.

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (iwsimon) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:20]: Duplicated "A new proxy contract and a new system contract to handle these transaction requests will be implemented".

**Review by iwsimon — DISMISSED** [review]
LGTM

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (lukelee-sl) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md]: fixed

**Review by iwsimon — DISMISSED** [review]
LGTM

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (david-bakin-sl) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:76]: "semantic correctness": does this mean `pureChecks` + `preHandle` before `handle` or what?  (probably could do with being explicit here in the design doc)

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (david-bakin-sl) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:77]: (presumably _this_ is `handle()`)

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (david-bakin-sl) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:87]: As a positive test this would be "Test that an EOA can call the `hbarApprove` function and grant an allowance to another account **when** the EOA signs the transaction".  The negative test for this is the first one in the "negative tests" section below.

Similarly for the second positive test here.

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (david-bakin-sl) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:90]: "another account or contract": or same account or contract too?

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (david-bakin-sl) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:95]: typo "posses" -> "possesses".  Or add more "s"s to suit you ....

**Review by david-bakin-sl — APPROVED** [review]
LGTM though there are some comments that could be addressed.

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (lukelee-sl) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:18]: We were trying to follow Ethereum EIP/ERC conventions.  For example the ERC20 standard which has an `IERC20` interface was introduced in EIP 20.  Since HAS was introduced in HIP 632, we followed the same pattern to come up with `IHRC632`

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (lukelee-sl) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:76]: I added a few more details but I don't think adding too many details about the `token service` is appropriate in this document.  The implementation is already finished and should probably be described in related token service documentation.

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (lukelee-sl) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:90]: this is a query and you can request allowance information for another account than the account making the request

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (Nana-EC) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:24]: Either that or we explore interface inheritance which I've been thinking on and say `IHRC906` and `IHRC632`
are separate interfaces and come together like this

```solidity
interface IHRC906 {
    function hbarAllowance(address owner, address spender) external returns (uint256 responseCode);
    function hbarApprove(address owner, address spender, int256 amount) external returns (uint256 responseCode);
}

interface IHRC632 {
    function isAuthorized(address account, bytes memory messageHash, bytes memory signatureBlob) external returns (bool authorizedResponse);
    function isAuthorizedRaw(address account, bytes memory messageHash, bytes memory signatureBlob) external returns (bool authorizedResponse);
}

abstract contract IHederaAccountService is IHRC906, IHRC632 {
}
```

Comment (david-bakin-sl) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:24]: Well, that was another thought I had - interface inheritance - but the reason I didn't suggest it was because the `IHederaAccountService` methods for `hbarAllowance` and `hbarApprove` have different arguments - and it isn't just an issue of overloading, which is fine, but the issue is who's doing the calling is different IIUC.  In the `IHederaAccountService` case it's some contract calling to get/set allowances on some other account, and in the `IHRC906` case it's us automagically calling the method on behalf of the EOA. Right?  (Or not?)

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (david-bakin-sl) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:106]: yes, I guess i'm suggesting that the positive tests include both doing it for yourself and for some other account

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (Nana-EC) [inline hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:69]: not sure if you meant this
```suggestion
The following pseudo code describes the logic during the `MessageFrame` construction for top level calls:
```

**Review by Nana-EC — APPROVED** [review]
LG, minor non blocking comments

File: hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md
Comment (Nana-EC) [reply hedera-node/docs/design/services/smart-contract-service/hbar-allowance-approval.md:24]: `IHRC906` is only automagically when called via an EOA directly. When called within contacts it's straight forward like the other system contract function and requires you be explicit about all the entities involved

**Review by Nana-EC — APPROVED** [review]

**Review by iwsimon — APPROVED** [review]
LGTM

**Pull request merged**
